New contributor workflow

dac09 opened the pull request:
Introduces new contributor workflow (without removing old `copy:watch` flow)

In order to contribute from a project now, you simply need to run

```yarn rwt link [RW_PATH]```

This will ask to setup your workspaces, and create the symlink for you. It willl also build and watch the redwood repo, so no need to build using another separate terminal session.

Once you're done, and you want to restore, you run

```yarn rwt restore```

### Need input
- [x] Is `rwt link` intuitive? Or should we call it something else? This is meant to be like npm link, yarn link
- [x] Improve the messaging, maybe use boxen to explain?
- [x] `rwt restore` -> `rwt unlink`? I don't like unlink for some reason
- [x] On my machine, in the redwood repo it creates a symlink inside packages. So packages/packages - no idea why!

### Other benefits

- [x] I think it consumes less cpu (one less watcher)
- [x] You can install / `yarn rw setup` stuff, without breaking the contribution workflow
- [x] Cleans up the redwood-tools.js file to make it easier to wrangle the code
- [x] Fixes infinite loop that sometimes happens building eslint-plugin
- [x] Fixes configs (webpack, etc.) not being available on the project until save. Since it's using symlinks, we don't have to worry about rebuilding!
- [x] We don't need to worry about manually installing new dependencies when using canary (just remember to yarn in your redwood repo) 

Preview of what it looks like:
https://s.tape.sh/NSEQV6FH

jtoar commented:
@dac09 super excited to try this out! Confirming the steps are:

0. create a new Redwood app and upgrade to this PR with `yarn rw upgrade --pr 1792:0.25.0-6b66c9f `
1. still in that redwood app, run `yarn rwt link [RW_PATH]`
2. say yes to the prompt
3. see edits from your local copy of the redwood repo live in the redwood app?
 
Here's a tape of me trying out a new setup command I'm working on. The repo doesn't seem to pick it up. Did I miss anything obvious? https://s.tape.sh/aIoTMMq2

dac09 commented:
Your process Looks fine to me Dom! Not sure why it's not coming through. Could you try:
a) yarn in the root of your project
b) Is there a soft link called "redwood" in the root of your project?

jtoar commented:
Tried `yarn` in the root and still no luck, but that soft link is there. Maybe it has to do with adding a new CLI command? I can try making changes to another part of the framework, like the router, and see if they propagate.

![image](https://user-images.githubusercontent.com/32992335/107864465-2e95ca00-6e11-11eb-805b-66dcbb13a965.png)

dac09 commented:
Weird. I've been using this to test itself (does that make any sense?!) I'll try it out more on Monday!

dac09 commented:
Ah, I can see why. When you build using the PR flow, it adds @rw/cli to your dev dependencies. For some reason this takes precedence over your locally linked packages. Once I removed this dependency form package.json and yarn'd, it all came through fine :)

jvanbaarsen commented:
@dac09 This is great! Would make the experience so much better :) Are there certain cases you'd like people to test?

jtoar commented:
@dac09 Those steps worked great! One thing I wanted to add: at the end you'll most likely have to make the `rw` binary executable again using `chmod +x ./node_modules/.bin/rw`.

jvanbaarsen commented:
@dac09 Just tested this locally works great! It's important to keep in mind the devDependency indeed (Would this be a problem when people upgraded their local repo with a PR upgrade?) One thing to note as @jtoar said, every time the thing did a rebuild you need to chmod the `rw` binary.

dac09 commented:
I definitely added the chmod inside the watcher... so everytime cli, apiserver and devserver builds, they get chmodded. I wonder why they hadn't in this case?

jvanbaarsen commented:
@dac09 You chmod dist/index.js, not the actual `rw` executable. Can that be the problem?

dac09 commented:
OH! I know why. It's because the changes to the build folder also need to be checked out in your local redwood repo LOL.

This is so meta!

jvanbaarsen commented:
@dac09 I was just about to post that 🙈 I'll double check this assumption with checkout out this PR in my RW reop.

dac09 commented:
Headsup 👒, I just rebased my branch. Now added `--clean` flag. 

@jvanbaarsen, @jtoar any chance you guys can see why 
> On my machine, in the redwood repo it creates a symlink inside packages. So packages/packages - no idea why!

jvanbaarsen commented:
@dac09 Your assumption was right! After checking out this PR as my RW base it works like a charm :) Also seeing the packages/packages thing though

jvanbaarsen commented:
@dac09 Can it be that because you're linking `packages` into `./redwood` and you then build again it kinda gets in a loop? So it keeps trying to create a folder packages inside of packages? (But since you cannot have a symlink in another symlink you just get one?)

dac09 commented:
I don't htink so @jvanbaarsen, the build is running in the redwood repo, not the actual project (cwd is set to frameworkPath). 

You can also do each of those steps manually, the function really isn't doing anything special.

jvanbaarsen commented:
@dac09 What is the reason `packages/*` is added to the project package.json?

jvanbaarsen commented:
@dac09 I got the problem I think :-) We need to add a check to see if the symlink is already present. If that is the case we should *not* create it. You can verify this assymption by removing the symlink, and then run this twice:

```
ln -sf YOUR_REDWOOD_REPO/packages ./redwood

```

jtoar commented:
Confirming that checking out this PR in my local redwood repo works!

jvanbaarsen commented on packages/cli/src/redwood-tools.js:
```diff
@@ -89,86 +103,232 @@ export const copyFiles = async (src, dest) => {
+
+  console.log(`Linking your local Redwood build from ${c.info(packagesPath)}`)
+
+  await execa(`ln -sf ${packagesPath} ./redwood`, {
+    shell: true,
+    stdio: 'inherit',
+    cwd: getPaths().base,
+  })
```
@dac09 This is the solution for the double packages/packages
```suggestion
  if (!fs.statSync('./redwood').isDirectory()) {
    await execa(`ln -sf ${packagesPath} ./redwood`, {
      shell: true,
      stdio: 'inherit',
      cwd: getPaths().base,
    })
  }

```

dac09 replied:
Nice!

thedavidprice commented:
This is looking 🤩

When it's ready, this likely needs either @Tobbe or @jeliasson (or both if possible) to run through on Windows.

dac09 commented:
I tried my best to keep it windows compatible! 🤞

dac09 commented:
We need to test a couple of other cases, before I move this PR out of draft

- [ ] Making changes to redwoodjs/core - are they reflected correctly in the project?

This is because core is a dev dependency of the root project, and as we've seen from the PR command, dev deps sometimes cause problems.

I'll check making changes to:

- [ ] auth
- [ ] router
- [ ] api

these are good cases, because they're included as dependencies of the web and api sides.

dac09 commented:
Tested cases mentioned in above comments. Note with change to api, you have to restart the dev server to see changes come through (but I don't think its related to this PR)

peterp reviewed (approved):
lgtm

jtoar commented:
Some thoughts on the things you asked for input on:
- `yarn rwt link` is intuitive but `yarn rwt restore` isn't as intuitive to me (i.e., that it's the opposite of link). I don't have an alternative for you though; `yarn rwt unlink` seemed fine to me? maybe `yarn rwt detach`? Also, [yarn doesn't have an unlink command](https://yarnpkg.com/cli/link/#details), fwiw
- why the prompt? as in, when would I say no to it? otherwise I figure let `yarn rwt link` do it's thing, unless you just want to make sure people know what they're getting into. here's the prompt for reference:
```terminal
You don't have redwood in your workspace package
? Add redwood folder to your workspaces config?
 This will modify your package.json › (y/N)
```
- @dthyresson mentioned this too, I tend to use the one-off `yarn rwt copy`, not `yarn rwt copy:watch`&mdash;does it seem worth it to add a non-watching version of `yarn rwt link`? or should I just embrace the watching workflow? any tips on how to do that would be greatly appreciated! 🙏

jvanbaarsen commented:
@jtoar Out of curiosity, why wouldn't you want to have the `watch` workflow? (Trying to understand the use case)

jtoar commented:
@jvanbaarsen probably just a limitation of the way I work, I'm used to making changes in one place then context switching to the other, so I don't feel like I need to `copy:watch`, I just `copy` when I'm ready. But the watch workflow's probably something I can adapt to. Do you do anything differently or is it really just work, save, switch?

dac09 commented:
I can add a flag so you can run `yarn rw l --watch false` ? @jtoar

Tobbe commented:
I ran into issues last time I tried cpw in Windows. Something about files being in use by another process. But when I tried yesterday with `link` I didn't get that error.

Should the flag be `--no-watch`?

dac09 commented:
Although I've added `no-xxx` as flags before, I think that its better than we just use a boolean to control these flags. With the defaults being true, for sensible default.

dac09 commented:
Btw - I've added to the added benefits section, I've found more cases where this can be useful!

jvanbaarsen commented:
@Tobbe I believe. if you have a boolean value for watch, you get `--no-watch` for free. 

@jtoar Yeah for me it is make a change in the Redwood repo, switch to test project and see if the change worked, the less steps the better for me :)

Tobbe commented:
> @Tobbe I believe. if you have a boolean value for watch, you get `--no-watch` for free.

Yeah, I think you're right, yargs does that, doesn't it? 👍

dac09 commented:
> @Tobbe I believe. if you have a boolean value for watch, you get --no-watch for free.

#TIL!

jvanbaarsen commented:
@Tobbe Yes! I actually got tripped up by this after someone created an issue with the --no-watch flag for the `rw test` command. I was about to open a PR to document that behaviour until I noticed it is something that is given for free by yargs.

jvanbaarsen commented:
@dac09 What is currently holding us back in merging this? Is there anything I can do to speed this up a bit more?

dac09 commented:
Just trying to minimise changes going into the release :)

dac09 commented:
@jvanbaarsen I also noticed this failing when I try it using canary. Something about the svgo binary not found (introduced in prerender image handling with Babel-plugin-inline-svg). Could you please see if you face the same problem?

jvanbaarsen commented:
@dac09 I was just about to test this and then realised I'm unsure what you mean with Canary? You either use Canary or the PR release right? There is no way to use both of them at the same time?

dac09 commented:
Hey @jvanbaarsen sorry for not being clear! I've just updated the branch (merging main into this). 

What we'll need to test is the latest PR build (once its updated).

dac09 commented:
@thedavidprice think this is ready for prime time. Shall we get it into 0.26? Docs are covered here: https://github.com/redwoodjs/redwood/pull/1803

thedavidprice reviewed (changes requested):
I'm not able to get this working using the instructions provided by @dac09 [in this comment](https://github.com/redwoodjs/redwood/pull/1792#issuecomment-779241812).

I can get the initial `yarn rwt link ...` to run. But not commands. I'm also seeing endless build:watch process:

https://user-images.githubusercontent.com/2951/108787306-55977e80-752a-11eb-8b50-ee7788b161e3.mov

Tobbe commented:
@thedavidprice I think you need to have your rw framework checked out to @dac09's branch (At least that's what I had to do to test this)

```
git checkout -b dac09-feature/contrib-workspaces main
git pull https://github.com/dac09/redwood.git feature/contrib-workspaces
```

dac09 commented:
@thedavidprice as Tobbe said, the infinite loop is caused by an existing bug in the build process. (Try this on any other branch, `yarn build:clean && yarn build:watch` in the redwood repo. The infinite loop is also fixed on this PR, so unless you have this branch checked out in your redwood repo, it'll infinite loop.
